internal/storage/prefixdb.go

```diff
@@ -0,0 +1,632 @@
+//
+// It implements the [Batch] interface for type prefixDBBatch.
+func (b prefixDBBatch) Set(key, value []byte) error {
+	if len(key) == 0 {
+		return errKeyEmpty
+	}
+	if value == nil {
+		return errValueNil
+	}
+
+	prefixedKey := prependPrefix(b.prefix, key)
+	if err := b.source.Set(prefixedKey, value); err != nil {
+		return fmt.Errorf("prefixed DB namespace batch set: %w", err)
+	}
+
+	return nil
+}
+
+// Delete adds a delete update to the batch that deletes database the entry for
+// key. It is safe to modify the contents of the arguments after Delete returns.
+//
+// It implements the [Batch] interface for type prefixDBBatch.
+func (b prefixDBBatch) Delete(key []byte) error {
+	if len(key) == 0 {
+		return errKeyEmpty
+	}
+
+	prefixedKey := prependPrefix(b.prefix, key)
+	if err := b.source.Delete(prefixedKey); err != nil {
+		return fmt.Errorf("prefixed DB namespace batch delete: %w", err)
+	}
+
+	return nil
+}
+
+// Write applies the batch to the database. Write does not guarantees that the batch
+// is persisted to disk before returning.
+//
+// It implements the [Batch] interface for type prefixDBBatch.
+func (b prefixDBBatch) Write() error {
+	if err := b.source.Write(); err != nil {
+		return fmt.Errorf("prefixed DB namespace batch write: %w", err)
+	}
+	return nil
+}
+
+// WriteSync applies the batch to the database. WriteSync guarantees that the batch
+// is persisted to disk before returning.
+//
+// It implements the [Batch] interface for type prefixDBBatch.
+func (b prefixDBBatch) WriteSync() error {
+	if err := b.source.WriteSync(); err != nil {
+		return fmt.Errorf("prefixed DB namespace batch write: %w", err)
+	}
+	return nil
+}
+
+// Close closes the batch without committing it. Close is idempotent, but calling
+// other methods on the batch after closing it will return an error.
+//
+// It implements the [Batch] interface for type prefixDBBatch.
+func (b prefixDBBatch) Close() error {
+	if err := b.source.Close(); err != nil {
+		return fmt.Errorf("prefixed DB namespace batch close: %w", err)
+	}
+	return nil
+}
+
+// prefixDBIterator wraps an [Iterator] interface representing an iterator iterating
+// over a range of key/value pairs in an underlying database. It will range over
+// keys that begin with the given prefix. That is, prefixDBIterator ensures that all
+// keys that it iterates over are scoped to the provided prefix.
+//
+// It is the caller's responsibility to call [Close] on it when done.
+// prefixDBIterator is not safe for concurrent use, but it is safe to use multiple
+// iterators concurrently. Callers cannot write to the underlying database
+// while there exists an iterator over it.
+// If there is an error during any operation, it is stored in the Iterator and can
+// be retrieved via the [Error] method.
+//
+// It implements the [Iterator] interface.
+type prefixDBIterator struct {
+	source Iterator
+	prefix []byte
+	start  []byte
+	end    []byte
+	valid  bool
+	err    error
```

Comment:
The `err` field is set only once in the `Valid()` method and is never returned elsewhere. All methods of `prefixDBIterator`, except `Domain()` and `Close()`,
will panic if `assertIsValid()` fails. While the `Error()` method does return `err`, is
it actually used if we panic most of the time?

What I'm asking is: Does it make sense to have this `err` field? Why doesn't `Valid()` also panic in that single case? It sets the iterator to invalid anyway, which will cause a panic on the next method call.

Reply:
I think the reason we're panicking is to enforce strict rules around `Iterator` usage.

Like you should always check if `Iterator` is valid before proceeding.

```go
...

it.Next()

if !it.Valid() {
  return it.Error()
}
```